### PR TITLE
Core/Scripts/DB: remove hardcoded text from npc_beaten_corpse

### PR DESCRIPTION
**Pull request by ghost**
**Changes proposed**:
- remove hardcoded text from script `npc_beaten_corpse`
- remove `OnGossipHello()` hook, gossip text moved to DB
- use `sGossipSelect()` hook instead of `OnGossipSelect()`
- add npc_text 3557 and 3558 as `text_id` to `gossip_menu`
- add gossip_menu_option 2871 to NPC 10668 Beaten Corpse
- add condition for `gossip_menu_option` 2871 in Quest ID 4921

Thanks to joschiwald for the sGossipSelect() script change.

**Target branch(es)**: 3.3.5

**Issues addressed**: Removes hardcoded text from the script `npc_beaten_corpse`

**Tests performed**: (Does it build, tested in-game, etc)
Compiles without errors, works as intended in-game.
Tested on 4 different horde characters/classes (horde quest), male & female.

**Known issues and TODO list**:
- [X] Replace `OnGossipSelect()` hook with `sGossipSelect()`

### CODE REVIEW

**Comment by joschiwald** [comment]
pls remove the gossip generating code since it is moved to db content, or move the whole npc script to sai

**Comment by tkrokli** [comment]
I have tried moving the script to SAI, but that can not be done unless we create a new SAI function having the same effect as `player->TalkedToCreature()`, maybe something like `SMART_ACTION_TALKED_TO_CREATURE`. This is because quest 4921 "Lost in Battle" does not have any external event or area explored setting that we currently can use to create a quest credit action in SAI. When I tried to use SAI action 15 `CALL_AREAEXPLOREDOREVENTHAPPENS`, I get an error message saying
```
SmartAIMgr: Entry 10668 SourceType 0 Event 1 Action 15 SpecialFlags for Quest entry 4921 does not include FLAGS_EXPLORATION_OR_EVENT(2), skipped.
```
To make this error message go away and make the SAI script work at all, I had to change this in the DB:
```sql
UPDATE `quest_template_addon` SET `SpecialFlags`= 2 WHERE `ID`= 4921;
```
But even if I implemented this in my DB, I only got "Objective completed", not any quest credit, which means it can only be completed by having the function `player->TalkedToCreature()` triggered.

You say you want me to remove the gossip generating code, do you mean I should remove the section `bool OnGossipHello(Player* player, Creature* creature) override` , or the lines 65, 74, 75, 76?

**Comment by tkrokli** [comment]
OK, so I have removed the gossip generating code as you asked, lines 63 + 74, but the result is that the gossip has stopped working completely. If there is a different way of doing it, I really want to know how.

**Comment by tkrokli** [comment]
OK, so I had to revert back to my original script version because it did not compile in Travis.
I still don't know how to follow up your request to "remove the gossip generating code" and still keep the gossip actions working as intended. I really hope someone can give me some pointers for that.

**Comment by joschiwald** [comment]
Remove gossiphello hook, change gossipselect hook to sgossipselect and remove the gossip stuff from there except the talkedtocreature line (i can show you an example in the evening)

File: src/server/scripts/Kalimdor/zone_the_barrens.cpp [hunk 1]
```diff
@@ -69,9 +71,9 @@ class npc_beaten_corpse : public CreatureScript
     bool OnGossipHello(Player* player, Creature* creature) override
     {
         if (player->GetQuestStatus(QUEST_LOST_IN_BATTLE) == QUEST_STATUS_INCOMPLETE || player->GetQuestStatus(QUEST_LOST_IN_BATTLE) == QUEST_STATUS_COMPLETE)
-            player->ADD_GOSSIP_ITEM(GOSSIP_ICON_CHAT, GOSSIP_CORPSE, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF+1);
+            player->ADD_GOSSIP_ITEM_DB(GOSSIP_MENU_OPTION_INSPECT_BODY, GOSSIP_OPTION_ID_BEATEN_CORPSE, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF + 1);
 
-        player->SEND_GOSSIP_MENU(3557, creature->GetGUID());
+        player->SEND_GOSSIP_MENU(NPC_TEXT_1_BEATEN_CORPSE, creature->GetGUID());
```
Comment (joschiwald) [inline src/server/scripts/Kalimdor/zone_the_barrens.cpp]:
```c++
class npc_beaten_corpse : public CreatureScript
{
    public:
        npc_beaten_corpse() : CreatureScript("npc_beaten_corpse") { }

        struct npc_beaten_corpseAI : public ScriptedAI
        {
            npc_beaten_corpseAI(Creature* creature) : ScriptedAI(creature)
            {
            }

            void sGossipSelect(Player* player, uint32 menuId, uint32 gossipListId) override
            {
                if (menuId == GOSSIP_MENU_OPTION_INSPECT_BODY && gossipListId == GOSSIP_OPTION_ID_BEATEN_CORPSE)
                {
                    player->CLOSE_GOSSIP_MENU();
                    player->TalkedToCreature(me->GetEntry(), me->GetGUID());
                }
            }
        };

        CreatureAI* GetAI(Creature* creature) const override
        {
            return new npc_beaten_corpseAI(creature);
        }
};
```

Comment (tkrokli) [inline src/server/scripts/Kalimdor/zone_the_barrens.cpp]: Oh, I see. I tried using only these lines
```cpp
            void sGossipSelect(Player* player, uint32 menuId, uint32 gossipListId) override
            {
                player->TalkedToCreature(me->GetEntry(), me->GetGUID());
            }
```
to replace the 2 existing hooks and of course it did not work. Thanks for the full example! :)

**Comment by tkrokli** [comment]
Thanks for the assistance, it works fine now that I got the help I needed with the `sGossipSelect()` part.

**Comment by tkrokli** [comment]
May I have some feedback if changes are needed, or can this PR be labeled [Feedback-PatchFix] ?

**Pull request merged**